File: server/server-core/src/main/java/com/simple/castle/server/screen/GameScreen.java
package com.simple.castle.server.screen;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;
import com.badlogic.gdx.graphics.g3d.Model;
import com.badlogic.gdx.graphics.g3d.ModelInstance;
import com.badlogic.gdx.math.Vector3;
import com.simple.castle.core.ModelSend;
import com.simple.castle.core.asset.AssetLoader;
import com.simple.castle.core.render.BaseCamera;
import com.simple.castle.core.render.BaseEnvironment;
import com.simple.castle.core.render.BaseRenderer;
import com.simple.castle.core.screen.BaseScreen;
import com.simple.castle.server.loader.SceneLoader;
import com.simple.castle.server.manager.SceneManager;
import com.simple.castle.server.physic.world.PhysicWorld;

import java.util.List;

public class GameScreen extends BaseScreen {

    private final Model model;
    private final SceneManager sceneManager;

    private final BaseRenderer baseRenderer;
    private final BaseEnvironment baseEnvironment;
    private final BaseCamera baseCamera;

    private final PhysicWorld physicWorld;

    private boolean isDebug = false;

    public GameScreen(BaseRenderer baseRenderer) {
        this.baseRenderer = baseRenderer;
        this.baseEnvironment = new BaseEnvironment();

        this.model = new AssetLoader().loadModel();
        this.sceneManager = new SceneManager(SceneLoader.loadSceneObjects(model), model);

        ModelInstance groundObject = sceneManager.getObject("ground").getModelInstance();
        Vector3 ground = groundObject.transform.getTranslation(new Vector3());
        this.baseCamera = new BaseCamera(ground, groundObject);

        this.physicWorld = new PhysicWorld();
        this.sceneManager.getAll().forEach(baseObject -> physicWorld.addRigidBody(baseObject.getPhysicObject()));

        this.inputMultiplexer.addProcessor(baseCamera);
    }

    @Override
    public void render(float delta) {
        baseCamera.update(delta);
        physicWorld.update(delta);
        baseRenderer.render(baseCamera, sceneManager.getDrawables(), baseEnvironment);

        if (isDebug) {
            physicWorld.debugDraw(baseCamera);
        }

    }

    @Override
    public void resize(int width, int height) {
        baseCamera.resize(width, height);
        baseCamera.update();
    }

    @Override
    public void dispose() {
        baseRenderer.dispose();
        physicWorld.dispose();
        sceneManager.dispose();
        model.dispose();
    }

    @Override
    public boolean keyDown(int keycode) {
        if (Input.Keys.ESCAPE == keycode) {
            Gdx.app.exit();
        }

        if (Input.Keys.F1 == keycode) {
            isDebug = !isDebug;
        }

        return super.keyDown(keycode);
    }

    public List<ModelSend> getState() {
        return sceneManager.getSend();
    }
}

File: server/server-core/src/main/java/com/simple/castle/server/kt/screen/GameScreen.kt
package com.simple.castle.server.kt.screen

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.graphics.g3d.Model
import com.badlogic.gdx.physics.bullet.collision.btCollisionObject
import com.simple.castle.core.kt.asset.AssetLoader
import com.simple.castle.core.kt.render.BaseCamera
import com.simple.castle.core.kt.render.BaseEnvironment
import com.simple.castle.core.kt.render.BaseRenderer
import com.simple.castle.core.kt.screen.BaseScreen
import com.simple.castle.server.kt.controller.PlayerController
import com.simple.castle.server.kt.loader.SceneLoader
import com.simple.castle.server.kt.manager.SceneManager
import com.simple.castle.server.kt.physic.CollisionEvent
import com.simple.castle.server.kt.physic.PhysicWorld

class GameScreen(private val baseRenderer: BaseRenderer) : BaseScreen(), CollisionEvent {
    private val model: Model = AssetLoader().loadModel()
    private val baseEnvironment: BaseEnvironment = BaseEnvironment()
    private val sceneLoader: SceneLoader = SceneLoader(model)
    private val sceneManager: SceneManager = SceneManager(sceneLoader.loadSceneObjects(), model)
    private val physicWorld: PhysicWorld = PhysicWorld(this as CollisionEvent, sceneManager.all.map { baseObject -> baseObject.physicObject })
    private val baseCamera: BaseCamera = BaseCamera(sceneManager.getObject("ground")?.modelInstance)
    private val playerController: PlayerController = PlayerController(sceneLoader.loadPlayers(), sceneManager, physicWorld)
    private var isDebug = false

    init {
        inputMultiplexer.addProcessor(baseCamera)
    }

    override fun render(delta: Float) {
        baseCamera.update(delta)
        physicWorld.update(delta)
        baseRenderer.render(baseCamera, sceneManager.drawables, baseEnvironment)
        if (isDebug) physicWorld.debugDraw(baseCamera)
    }

    override fun resize(width: Int, height: Int) {
        baseCamera.resize(width, height)
        baseCamera.update()
    }

    override fun dispose() {
        baseRenderer.dispose()
        physicWorld.dispose()
        sceneManager.dispose()
        model.dispose()
    }

    override fun keyDown(keycode: Int): Boolean {
        if (Input.Keys.ESCAPE == keycode) {
            Gdx.app.exit()
        }
        if (Input.Keys.F1 == keycode) {
            isDebug = !isDebug
        }
        if (Input.Keys.SPACE == keycode) {
            playerController.spawn()
        }
        return super.keyDown(keycode)
    }

    override fun onContactStarted(colObj0: btCollisionObject, colObj1: btCollisionObject) {
        playerController.onContactStarted(colObj0, colObj1)
    }

}